tamagotchi/src/contract.rs

```diff
@@ -0,0 +1,182 @@
+
+    debug_assert!(state.is_some(), "state isn't initialized");
+
+    unsafe { state.unwrap_unchecked() }
+}
+
+#[no_mangle]
+unsafe extern "C" fn init() {
+    let tamagotchi_name: String = msg::load().expect("Error in init message");
+    let new_tamagotchi: Tamagotchi = Tamagotchi {
+        name: tamagotchi_name,
+        date_of_birth: exec::block_timestamp(),
+        owner: msg::source(),
+        fed: 5000,
+        fed_block: 0,
+        entertained: 5000,
+        entertained_block: 0,
+        rested: 5000,
+        rested_block: 0
+    };
+    TAMAGOTCHI = Some(new_tamagotchi);
+    msg::reply("successful initialization!", 0).expect("error in reply");
+}
+
+#[no_mangle]
+unsafe extern "C" fn handle() {
+    let type_message: TmgAction = msg::load().expect("error in load message");
+    let tamagotchi = state_mut();
+
+    let blocksHeight = blocks_height();
+
+    match type_message {
+        TmgAction::Name => {
+            let tamagotchiName: TmgEvent = TmgEvent::Name(String::from(&tamagotchi.name));
+            msg::reply(tamagotchiName, 0).expect("Error in sending tamagotchi name");
+        },
+        TmgAction::Age => {
+            let tamagotchiAge: u64 = exec::block_timestamp();
+            tamagotchi.date_of_birth = tamagotchiAge;
+            msg::reply(TmgEvent::Age(tamagotchiAge), 0).expect("Errorin sending tamagotchi age");
+        },
+        TmgAction::Feed => {
+            tamagotchi.fed = tamagotchi_fed_updated(blocksHeight);
+            tamagotchi.fed_block = blocksHeight;
+            msg::reply(TmgEvent::Fed, 0).expect("Error sending tamagotchi variant 'Fed'");
+        },
+        TmgAction::Play => {
+            tamagotchi.entertained = tamagotchi_entertained_updated(blocksHeight);
+            tamagotchi.entertained_block = blocksHeight;
+            msg::reply(TmgEvent::Entertained, 0).expect("Error sending tamagotchi variant 'Entertained'");  
+        },
+        TmgAction::Sleep => {
+            tamagotchi.rested = tamagotchi_rested_updated(blocksHeight);
+            tamagotchi.rested_block = blocksHeight;
+            msg::reply(TmgEvent::Slept, 0).expect("Error sending tamagotchi variant 'Slept'");  
+        }
+    }
+}
+
+#[no_mangle]
+extern "C" fn state() {
+    let blocksHeight = blocks_height();
+    let tamagotchi = state_ref();
+    let state = Tamagotchi {
+        name: String::from(&tamagotchi.name),
+        date_of_birth: tamagotchi.date_of_birth,
+        owner: tamagotchi.owner.clone(),
+        fed: tamagotchi_total_fed(blocksHeight),
+        fed_block: tamagotchi.fed_block,
+        entertained: tamagotchi_total_entertained(blocksHeight),
+        entertained_block: tamagotchi.entertained_block,
+        rested: tamagotchi_total_rested(blocksHeight),
+        rested_block: tamagotchi.rested_block
+    };
+    msg::reply(tamagotchi, 0).expect("Failed to share state");
+}
+
+fn blocks_height() -> u64 {
+    u64::from(exec::block_height())
```

Comment:
Usually, we use `as` for ordinal types.
```suggestion
    exec::block_height() as u64
```